contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,749 @@
+command, `kubectl debug`, which parallels an existing command, `kubectl exec`.
+Whereas `kubectl exec` runs a *process* in a *container*, `kubectl debug` will
+be similar but run a *container* in a *pod*.
+
+A container created by `kubectl debug` is a *Debug Container*. Just like a
+process run by `kubectl exec`, a Debug Container is not part of the pod spec and
+has no resource stored in the API. Unlike `kubectl exec`, a Debug Container
+*does* have status that is reported in `v1.PodStatus` and displayed by `kubectl
+describe pod`.
+
+For example, the following command would attach to a newly created container in
+a pod:
+
+```
+kubectl debug -c debug-shell --image=debian target-pod -- bash
+```
+
+It would be reasonable for Kubernetes to provide a default container name and
+image, making the minimal possible debug command:
+
+```
+kubectl debug target-pod
+```
+
+This creates an interactive shell in a pod which can examine and signal other
+processes in the pod. It has access to the same network and IPC as processes in
+the pod. It can access the filesystem of other processes by `/proc/$PID/root`.
+As is already the case with regular containers, Debug Containers can enter
+arbitrary namespaces of another container via `nsenter` when run with
+`CAP_SYS_ADMIN`.
+
+*Please see the User Stories section for additional examples and Alternatives
+Considered for the considerable list of other solutions we considered.*
+
+## Implementation Details
+
+The implementation of `kubectl debug` closely mirrors the implementation of
+`kubectl exec`, with most of the complexity implemented in the `kubelet`. How
+functionality like this best fits into Kubernetes API has been contentious. In
+order to make progress, we will start with the smallest possible API change,
+extending `/exec` to support Debug Containers, and iterate.
+
+From the perspective of the user, there's a new command, `kubectl debug`, that
+creates a Debug Container and attaches to its console. We believe a new command
+will be less confusing for users than overloading `kubectl exec` with a new
+concept. Users give Debug Containers a name (e.g. "debug" or "shell") which can
+subsequently be used to reattach and is reported by `kubectl describe`.
+
+### Kubernetes API Changes
+
+#### Chosen Solution: "exec++"
+
+We will extend `v1.Pod`'s `/exec` subresource to support "executing" container
+images. The current `/exec` endpoint must implement `GET` to support streaming
+for all clients. We don't want to encode a (potentially large) `v1.Container` as
+an HTTP parameter, so we must extend `v1.PodExecOptions` with the specific
+fields required for creating a Debug Container:
+
+```
+// PodExecOptions is the query options to a Pod's remote exec call
+type PodExecOptions struct {
+        ...
+        // Run Command in an ephemeral container which shares some namespaces with Container.
+        EphemeralContainer PodExecEphemeralContainerSpec
```

Comment:
This needs to be a pointer. Uh, but I will suggest something else since sub structs are indeed currently broken.

Reply:
```go
// PodExecOptions is the query options to a Pod's remote exec call
type PodExecOptions struct {
  ...
  // EphemeralContainerName is the name of an ephemeral container in which the
  // command ought to be run. Either both EphemeralContainerName and
  // EphemeralContainerImage fields must be set, or neither.
  EphemeralContainerName *string `json:"ephemeralContainerName,omitempty" ...`

  // EphemeralContainerImage is the image of an ephemeral container in which the command
  // ought to be run. Either both EphemeralContainerName and EphemeralContainerImage
  // fields must be set, or neither.
  EphemeralContainerImage *string `json:"ephemeralContainerImage,omitempty" ...`
}
```

Reply:
Renamed as suggested